compiler/rustc_session/src/filesearch.rs

```diff
@@ -123,15 +123,44 @@ pub fn get_or_default_sysroot() -> PathBuf {
         fix_windows_verbatim_for_gcc(&path)
     }
 
-    match env::current_exe() {
-        Ok(exe) => {
-            let mut p = canonicalize(exe);
-            p.pop();
-            p.pop();
-            p
+    // Use env::current_exe() to get the path of the executable following
+    // symlinks/canonicalizing components.
+    fn from_current_exe() -> PathBuf {
+        match env::current_exe() {
+            Ok(exe) => {
+                let mut p = canonicalize(exe);
+                p.pop();
+                p.pop();
+                p
+            }
+            Err(e) => panic!("failed to get current_exe: {}", e),
+        }
+    }
+
+    // Use env::args().next() instead of env::current_exe() to get the path of
+    // the executable without following symlinks/canonicalizing any component.
+    // This makes the rustc binary able to locate Rust libraries in systems
+    // using content-addressable storage (CAS). However, since it can be set to
+    // an arbitrary value, fs::get_or_default_sysroot() and its return value
+    // should not be relied on for security-related decisions.
```

Comment:
This last sentence seems misleading, as it applies just as much to from_current_exe. I would perhaps expect to see it on the get_or_default_sysroot function itself.

Reply:
Done. (I improved the changes a bit and fixed the comments.)